File: app/src/main/java/com/chess/cryptobot/api/LivecoinMarketService.java
package com.chess.cryptobot.api;

import com.chess.cryptobot.model.response.livecoin.LivecoinAddressResponse;
import com.chess.cryptobot.model.response.livecoin.LivecoinBalanceResponse;
import com.chess.cryptobot.model.response.livecoin.LivecoinCurrenciesListResponse;
import com.chess.cryptobot.model.response.livecoin.LivecoinOrderBookResponse;
import com.chess.cryptobot.model.response.livecoin.LivecoinResponse;
import com.chess.cryptobot.model.response.livecoin.LivecoinTickerResponse;
import com.chess.cryptobot.model.response.livecoin.LivecoinTradeLimitResponse;

import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.HeaderMap;
import retrofit2.http.POST;
import retrofit2.http.QueryMap;

public interface LivecoinMarketService {

    @GET("payment/balance")
    Call<LivecoinBalanceResponse> getBalance(@QueryMap(encoded = true) Map<String, String> options, @HeaderMap Map<String, String> headers);

    @GET("exchange/order_book")
    Call<LivecoinOrderBookResponse> getOrderBook(@QueryMap(encoded = true) Map<String, String> options);

    @GET("exchange/ticker")
    Call<List<LivecoinTickerResponse>> getTicker();

    @GET("info/coinInfo")
    Call<LivecoinCurrenciesListResponse> getCurrencies();

    @GET("exchange/restrictions")
    Call<LivecoinTradeLimitResponse> getMinTradeSize();

    @GET("payment/get/address")
    Call<LivecoinAddressResponse> getAddress(@QueryMap(encoded = true) Map<String, String> options, @HeaderMap Map<String, String> headers);

    @FormUrlEncoded
    @POST("payment/out/coin")
    Call<LivecoinResponse> payment(@Field(value = "amount", encoded = true) String amount,
                                   @Field(value = "currency", encoded = true) String currency,
                                   @Field(value = "wallet", encoded = true) String wallet,
                                   @HeaderMap Map<String, String> headers);

    @FormUrlEncoded
    @POST("exchange/buylimit")
    Call<LivecoinResponse> buy(@Field(value = "currencyPair", encoded = true) String currencyPair,
                                    @Field(value = "price", encoded = true) String price,
                                    @Field(value = "quantity", encoded = true) String quantity,
                                    @HeaderMap Map<String, String> headers);

    @FormUrlEncoded
    @POST("exchange/selllimit")
    Call<LivecoinResponse> sell(@Field(value = "currencyPair", encoded = true) String currencyPair,
                                     @Field(value = "price", encoded = true) String price,
                                     @Field(value = "quantity", encoded = true) String quantity,
                                     @HeaderMap Map<String, String> headers);
}

File: app/src/main/java/com/chess/cryptobot/api/LivecoinMarketService.kt
package com.chess.cryptobot.api

import com.chess.cryptobot.model.response.livecoin.LivecoinAddressResponse
import com.chess.cryptobot.model.response.livecoin.LivecoinBalanceResponse
import com.chess.cryptobot.model.response.livecoin.LivecoinCurrenciesListResponse
import com.chess.cryptobot.model.response.livecoin.LivecoinHistoryResponse
import com.chess.cryptobot.model.response.livecoin.LivecoinOrderBookResponse
import com.chess.cryptobot.model.response.livecoin.LivecoinOrdersResponse
import com.chess.cryptobot.model.response.livecoin.LivecoinResponse
import com.chess.cryptobot.model.response.livecoin.LivecoinTickerResponse
import com.chess.cryptobot.model.response.livecoin.LivecoinTradeLimitResponse
import retrofit2.Call
import retrofit2.http.*

interface LivecoinMarketService {
    @GET("payment/balance")
    fun getBalance(@QueryMap(encoded = true) options: Map<String, String>?, @HeaderMap headers: Map<String, String>?): Call<LivecoinBalanceResponse?>?

    @GET("exchange/order_book")
    fun getOrderBook(@QueryMap(encoded = true) options: Map<String, String>): Call<LivecoinOrderBookResponse>

    @get:GET("exchange/ticker")
    val ticker: Call<List<LivecoinTickerResponse>>

    @get:GET("info/coinInfo")
    val currencies: Call<LivecoinCurrenciesListResponse>

    @get:GET("exchange/restrictions")
    val minTradeSize: Call<LivecoinTradeLimitResponse>

    @GET("payment/get/address")
    fun getAddress(@QueryMap(encoded = true) options: Map<String, String>?, @HeaderMap headers: Map<String, String>?): Call<LivecoinAddressResponse?>?

    @FormUrlEncoded
    @POST("payment/out/coin")
    fun payment(@Field(value = "amount", encoded = true) amount: String?,
                @Field(value = "currency", encoded = true) currency: String?,
                @Field(value = "wallet", encoded = true) wallet: String?,
                @HeaderMap headers: Map<String?, String?>?): Call<LivecoinResponse?>?

    @FormUrlEncoded
    @POST("exchange/buylimit")
    fun buy(@Field(value = "currencyPair", encoded = true) currencyPair: String?,
            @Field(value = "price", encoded = true) price: String?,
            @Field(value = "quantity", encoded = true) quantity: String?,
            @HeaderMap headers: Map<String, String>?): Call<LivecoinResponse?>?

    @FormUrlEncoded
    @POST("exchange/selllimit")
    fun sell(@Field(value = "currencyPair", encoded = true) currencyPair: String?,
             @Field(value = "price", encoded = true) price: String?,
             @Field(value = "quantity", encoded = true) quantity: String?,
             @HeaderMap headers: Map<String, String>?): Call<LivecoinResponse?>?

    @GET("payment/history/transactions")
    fun getHistory(@QueryMap(encoded = true) options: Map<String, String>?, @HeaderMap headers: Map<String, String>): Call<List<LivecoinHistoryResponse>>

    @GET("exchange/client_orders")
    fun getOpenOrders(@QueryMap(encoded = true) options: Map<String, String>?, @HeaderMap headers: Map<String, String>?): Call<LivecoinOrdersResponse?>?
}